### DISCUSSION TITLE
What kind of training scores did you get while training?

### DISCUSSION THREAD

**Discussion by JefeDryden** [DonsetPG/narya#17] [Q&A] [answered]
Hi all,

I've been working to try and possibly adapt this code for keypoint detection use in a Hockey setting. I've created a dataset of about 260 images (small, I know) and trained/tested a couple of models. When it works, it's great, but so far still very inaccurate. I was wondering what kind of metrics you received while training the model to possibly see how far off I am and how much more data I need to collect. 

I've had an absolute delight working through your code and trying to make it work in a hockey setting, thanks so much for sharing it with the world. I've attached some pictures of the model to show the first few fruits of my labour! 

![HockeyDetection2](https://user-images.githubusercontent.com/55153697/107913683-d94fda80-6f2e-11eb-8411-5fff12fc0631.PNG)
![Hockey Keypoint Detection1](https://user-images.githubusercontent.com/55153697/107913686-d94fda80-6f2e-11eb-8f11-dd7ee87e82ce.PNG)

**Comment by DonsetPG** [comment, marked as answer]
Hey, 

Thanks for the message! 

Here are the results we obtained while training the KeypointDetection model: 

<img width="613" alt="ablaKp" src="https://user-images.githubusercontent.com/44346877/107920063-9ca2d400-6f6c-11eb-93e7-beb8c673d087.png">

and here the results we obtained when computing the homography with those keypoints:

<img width="734" alt="Capture d’écran 2021-02-15 à 08 38 25" src="https://user-images.githubusercontent.com/44346877/107920150-c2c87400-6f6c-11eb-91a3-c7b4a0d50d93.png">

You can find results related to Hockey [here](https://openaccess.thecvf.com/content_WACV_2020/papers/Jiang_Optimizing_Through_Learned_Errors_for_Accurate_Sports_Field_Registration_WACV_2020_paper.pdf) and [here](https://openaccess.thecvf.com/content/WACV2021/papers/Nie_A_Robust_and_Efficient_Framework_for_Sports-Field_Registration_WACV_2021_paper.pdf).

I think you can already try the upgrade from [this paper](https://openaccess.thecvf.com/content/WACV2021/papers/Nie_A_Robust_and_Efficient_Framework_for_Sports-Field_Registration_WACV_2021_paper.pdf), using a uniform grid for the keypoints. 

I hope that this is useful.

**Comment by DonsetPG** [comment]
Hey @JefeDryden, I was wondering if you were able to obtain any results on hockey datasets? 
If not, let me know if we can help!
Thanks a lot

**Comment by JefeDryden** [comment]
Hi @DonsetPG ,

I actually did. It turned out in my original training attempt that the numbers on my training picture labels compared to the annotation data had shifted by 3... felt pretty embarrassed figuring that one out. 

As for my training results on actual data, at my best I was able to train a model which picked up 30/32 keypoints for a hockey rink perfectly, (Testing F1 score of around 0.95) and I was finding that at random the model seemed to train and I guess end up in a different local minima each time. I'm not sure if you noticed something like that while you were training. I ended up spending a week with two different colab notebooks open continually running training programs to try and get the best model. I imagine with more data I can achieve better results. I made about 450 annotated pictures for training data.

As for actual results, you can see them here! I found a way to combine your keypoint detector and homography generator with a YoloV4 model I had made previously which gave me these results here (3 separate videos)

Video 1: Determining the homography of the template image onto the screen: https://www.youtube.com/watch?v=Ab1BaHRJ1hs
Video 2: Warping the Broadcast footage into a bird's eye view into the rink: https://www.youtube.com/watch?v=LOz-x0NJfV8
Video 3: Combining the detections made in YoloV4 with the keypoint homography detector: https://www.youtube.com/watch?v=KaBG52week4

I've been really happy with the results, and it's decently quick in a Google Colab Pro notebook, was able to detect and create the video at around 7 FPS. 

In terms of moving forward, I'd love to be able to figure out how you made the homography dataset as opposed to the keypoint one. I haven't quite been able to figure that out but I have a feeling that it could be a good next step for me. Otherwise it would be attempting to figure out the grid keypoint system that you have been talking about in the other discussion board.

Thanks so much again!

**Reply by DonsetPG** [reply]
Hey, thanks for the answer. 

> I'm not sure if you noticed something like that while you were training. I ended up spending a week with two different colab notebooks open continually running training programs to try and get the best model. I imagine with more data I can achieve better results. I made about 450 annotated pictures for training data.

I don't recall having this kind of issue with the key point model (but definitively with the homography model haha)

The results looks pretty good, congrats! 

> In terms of moving forward, I'd love to be able to figure out how you made the homography dataset as opposed to the keypoint one. I haven't quite been able to figure that out but I have a feeling that it could be a good next step for me. Otherwise it would be attempting to figure out the grid keypoint system that you have been talking about in the other discussion board.

The homography dataset comes from another paper (back in 2017 If I remember correctly). I have no informations about they built it, but looking at the original dataset it may have been with Matlab. 
Looking at recent papers, I think that the direct homography prediction is actually pretty difficult, and that other approach (grid key points, features detections, etc) seems more promising. 

__ 

I don't know if your are willing to share your datasets / and-or model, but if yes, please feel free to make a pull request! (or link the data and I'll make the modifications). It would be great if the repo can start to add new sports! 

(Edit: @JefeDryden : weights/datasets and a Colab would be great, I'll then be able to add it to the current library (you'll get credited in the code and in the readme of course!))

Have a nice weekend

**Reply by jbrumwell** [reply]
@JefeDryden this looks great well done, I am working on a project and would love to test this out :)

**Comment by DonsetPG** [comment]
Hey @JefeDryden , just wanted to know if you had some time to have a look at my last message? Otherwise no worries. 
Have a nice week end, cheers
